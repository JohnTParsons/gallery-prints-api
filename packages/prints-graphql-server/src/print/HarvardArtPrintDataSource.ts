import { RESTDataSource } from 'apollo-datasource-rest';

// FIXME - throw error, rather than using fallbacks, if env var(s) not set?
const HARVARD_ART_BASE_URL = process.env.HARVARD_ART_BASE_URL || 'https://api.harvardartmuseums.org/';
const HARVARD_ART_API_KEY = process.env.HARVARD_ART_API_KEY || 'c28e4be0-4c0e-11ea-90d6-25d9a9fe80fc';

export class HarvardArtPrintDataSource extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = HARVARD_ART_BASE_URL;
  }

  async getPrints(sortdir = 'DESC', pagenumber = 1, pagesize = 10) {
    console.log(`Calling getPrints with pagesize=${pagesize} pagenumber=${pagenumber}`);
    const { info, records } = await this.get(
      `object?apikey=${HARVARD_ART_API_KEY}&classification=Prints&size=${pagesize}&page=${pagenumber}`,
    );
    return {
      pageinfo: { pagenumber, totalpages: Math.ceil(info.totalrecords / pagesize), totalrecords: info.totalrecords },
      prints: records
        .map(({ id, title, technique, primaryimageurl, rank }) => ({ id, title, technique, primaryimageurl, rank }))
        .filter((record) => record.primaryimageurl && record)
        .sort((a, b) => (sortdir === 'DESC' ? b.rank - a.rank : a.rank - b.rank)),
    };
  }
}
