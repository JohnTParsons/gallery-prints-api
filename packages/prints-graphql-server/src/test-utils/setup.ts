import { RESTDataSource } from 'apollo-datasource-rest';
import { InMemoryLRUCache } from 'apollo-server-caching';

/**
 * Initialize the service.
 * NOTE:
 * ==================================
 * We need to initialize service (dataSource) manually as if it's used through apollo server, apollo server also do the same.
 * ==================================
 * @param Service
 */
export const initialize = <T extends RESTDataSource>(Service: new () => T): T => {
  const service = new Service();

  service.initialize({
    context: {},
    cache: new InMemoryLRUCache(),
  });
  return service;
};
