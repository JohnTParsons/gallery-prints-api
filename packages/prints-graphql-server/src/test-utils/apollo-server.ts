import { createTestClient } from 'apollo-server-testing';
//import schema from '../schema'; // import existing schema & datasource to utilize the same in test scenarios as well...
import { ApolloServer, gql } from 'apollo-server';
import { RandomUserDataSource } from '../person/RandomUserDataSource';
import { HarvardArtPrintDataSource } from '../print/HarvardArtPrintDataSource';
import * as commonTypes from '../common/common-types';
import * as personTypes from '../person/person-types';
import * as printTypes from '../print/print-types';
import * as personResolvers from '../person/person-resolvers';
import * as printResolvers from '../print/print-resolvers';

const typeDefs = gql`
  ${commonTypes.typeDefString}
  ${personTypes.typeDefString}
  ${printTypes.typeDefString}
  type Query {
    ${personTypes.queryTypeDefString}
    ${printTypes.queryTypeDefString}
  }
`;

const resolvers = {
  Query: {
    ...personResolvers.resolvers.Query,
    ...printResolvers.resolvers.Query,
  },
};

const createApolloTestServer = () => {
  const server = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({
      randomUserAPI: new RandomUserDataSource(),
      harvardArtPrintAPI: new HarvardArtPrintDataSource(),
    }),
    debug: true,
    context: () => {
      // Important: request context is not available during jest testing, hence mocked the request context here.
      return { token: 'foo' };
    },
  });
  return server;
};

const apolloServer = createApolloTestServer();

// @ts-ignore
export const { query, mutate } = createTestClient(apolloServer);
