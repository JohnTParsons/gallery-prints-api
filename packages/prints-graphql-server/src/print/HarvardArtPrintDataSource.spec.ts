import nock from 'nock';
import { initialize } from '../test-utils/setup';
import { HarvardArtPrintDataSource } from './HarvardArtPrintDataSource';
import restPrints from '../../../../fixtures/print/rest_printsonpage.json';

describe('HarvardArtPrintDataSource', () => {
  let harvardArtPrintAPI: HarvardArtPrintDataSource;

  beforeAll(() => {
    harvardArtPrintAPI = initialize(HarvardArtPrintDataSource);
  });

  it('by default, should return prints on page 1 in descending order', async () => {
    nock(
      'https://api.harvardartmuseums.org',
      // Wildcard match the path for simplicity
      {
        filteringScope: function (scope) {
          return true;
        },
      },
    )
      .filteringPath(function (path) {
        return '/';
      })
      .get('/')
      .reply(200, restPrints);
    const response = await harvardArtPrintAPI.getPrints();
    expect(response.pageinfo).toEqual({ pagenumber: 1, totalpages: 6878, totalrecords: 68777 });
    expect(response.prints).toHaveLength(9); // not 10 because there is an empty image we are filtering out
    expect(response.prints[0].rank).toBeGreaterThan(response.prints[8].rank);
    expect(response).toMatchSnapshot();
  });

  it('should return prints on page 1 in ascending order when given that sortdir', async () => {
    nock(
      'https://api.harvardartmuseums.org',
      // Wildcard match the path for simplicity
      {
        filteringScope: function (scope) {
          return true;
        },
      },
    )
      .filteringPath(function (path) {
        return '/';
      })
      .get('/')
      .reply(200, restPrints);
    const response = await harvardArtPrintAPI.getPrints('ASC');
    expect(response.pageinfo).toEqual({ pagenumber: 1, totalpages: 6878, totalrecords: 68777 });
    expect(response.prints).toHaveLength(9); // not 10 because there is an empty image we are filtering out
    expect(response.prints[0].rank).toBeLessThan(response.prints[8].rank);
    expect(response).toMatchSnapshot();
  });

  it('should return prints on page 2 in ascending order when given that sortdir', async () => {
    nock(
      'https://api.harvardartmuseums.org',
      // Wildcard match the path for simplicity
      {
        filteringScope: function (scope) {
          return true;
        },
      },
    )
      .filteringPath(function (path) {
        return '/';
      })
      .get('/')
      .reply(200, restPrints);
    const response = await harvardArtPrintAPI.getPrints('ASC', 2);
    expect(response.pageinfo).toEqual({ pagenumber: 2, totalpages: 6878, totalrecords: 68777 });
    expect(response.prints).toHaveLength(9); // not 10 because there is an empty image we are filtering out
    expect(response.prints[0].rank).toBeLessThan(response.prints[8].rank);
    expect(response).toMatchSnapshot();
  });
});
