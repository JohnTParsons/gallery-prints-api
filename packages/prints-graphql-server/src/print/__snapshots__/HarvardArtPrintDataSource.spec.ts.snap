// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HarvardArtPrintDataSource by default, should return prints on page 1 in descending order 1`] = `
Object {
  "pageinfo": Object {
    "pagenumber": 1,
    "totalpages": 6878,
    "totalrecords": 68777,
  },
  "prints": Array [
    Object {
      "id": 77295,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:DDC107153_dynmc",
      "rank": 267351,
      "technique": "Etching",
      "title": "Nonce Palace at San Francesco della Vigna in Venice",
    },
    Object {
      "id": 77297,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:DDC107129_dynmc",
      "rank": 267328,
      "technique": "Etching",
      "title": "Horse-Drawn Carriage with Well-Dressed Figures and Bathers; Fortress and Port in the Background",
    },
    Object {
      "id": 77304,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:DDC107121_dynmc",
      "rank": 267320,
      "technique": "Etching",
      "title": "Viceroy's Arrival at the Pleasure-Garden in Naples",
    },
    Object {
      "id": 77264,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:DDC107120_dynmc",
      "rank": 267319,
      "technique": "Etching",
      "title": "Arrival in Naples at the Piazza di Spagna",
    },
    Object {
      "id": 76522,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV048179_dynmc",
      "rank": 236028,
      "technique": "Lithograph",
      "title": "Twentysix Gasoline Stations",
    },
    Object {
      "id": 76497,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV044654_dynmc",
      "rank": 236006,
      "technique": "Etching",
      "title": "The Drawing School of Francesco Rosapina by Lamplight",
    },
    Object {
      "id": 77466,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV017572_dynmc",
      "rank": 57509,
      "technique": null,
      "title": "Triptych: Nambyō Ryōji",
    },
    Object {
      "id": 77248,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:CARP06376_dynmc",
      "rank": 57490,
      "technique": null,
      "title": "Triptych: Actors as the Sixteen Arhats (Mitate: Jūroku Rakan)",
    },
    Object {
      "id": 76963,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:CARP06289_dynmc",
      "rank": 54215,
      "technique": null,
      "title": "Triptych: Seven Gods of Good Fortune (Shichifukujin) Playing Music and Dancing in the Ship of Treasures (Takarabune)",
    },
  ],
}
`;

exports[`HarvardArtPrintDataSource should return prints on page 1 in ascending order when given that sortdir 1`] = `
Object {
  "pageinfo": Object {
    "pagenumber": 1,
    "totalpages": 6878,
    "totalrecords": 68777,
  },
  "prints": Array [
    Object {
      "id": 76963,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:CARP06289_dynmc",
      "rank": 54215,
      "technique": null,
      "title": "Triptych: Seven Gods of Good Fortune (Shichifukujin) Playing Music and Dancing in the Ship of Treasures (Takarabune)",
    },
    Object {
      "id": 77248,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:CARP06376_dynmc",
      "rank": 57490,
      "technique": null,
      "title": "Triptych: Actors as the Sixteen Arhats (Mitate: Jūroku Rakan)",
    },
    Object {
      "id": 77466,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV017572_dynmc",
      "rank": 57509,
      "technique": null,
      "title": "Triptych: Nambyō Ryōji",
    },
    Object {
      "id": 76497,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV044654_dynmc",
      "rank": 236006,
      "technique": "Etching",
      "title": "The Drawing School of Francesco Rosapina by Lamplight",
    },
    Object {
      "id": 76522,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV048179_dynmc",
      "rank": 236028,
      "technique": "Lithograph",
      "title": "Twentysix Gasoline Stations",
    },
    Object {
      "id": 77264,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:DDC107120_dynmc",
      "rank": 267319,
      "technique": "Etching",
      "title": "Arrival in Naples at the Piazza di Spagna",
    },
    Object {
      "id": 77304,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:DDC107121_dynmc",
      "rank": 267320,
      "technique": "Etching",
      "title": "Viceroy's Arrival at the Pleasure-Garden in Naples",
    },
    Object {
      "id": 77297,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:DDC107129_dynmc",
      "rank": 267328,
      "technique": "Etching",
      "title": "Horse-Drawn Carriage with Well-Dressed Figures and Bathers; Fortress and Port in the Background",
    },
    Object {
      "id": 77295,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:DDC107153_dynmc",
      "rank": 267351,
      "technique": "Etching",
      "title": "Nonce Palace at San Francesco della Vigna in Venice",
    },
  ],
}
`;

exports[`HarvardArtPrintDataSource should return prints on page 2 in ascending order when given that sortdir 1`] = `
Object {
  "pageinfo": Object {
    "pagenumber": 2,
    "totalpages": 6878,
    "totalrecords": 68777,
  },
  "prints": Array [
    Object {
      "id": 76963,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:CARP06289_dynmc",
      "rank": 54215,
      "technique": null,
      "title": "Triptych: Seven Gods of Good Fortune (Shichifukujin) Playing Music and Dancing in the Ship of Treasures (Takarabune)",
    },
    Object {
      "id": 77248,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:CARP06376_dynmc",
      "rank": 57490,
      "technique": null,
      "title": "Triptych: Actors as the Sixteen Arhats (Mitate: Jūroku Rakan)",
    },
    Object {
      "id": 77466,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV017572_dynmc",
      "rank": 57509,
      "technique": null,
      "title": "Triptych: Nambyō Ryōji",
    },
    Object {
      "id": 76497,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV044654_dynmc",
      "rank": 236006,
      "technique": "Etching",
      "title": "The Drawing School of Francesco Rosapina by Lamplight",
    },
    Object {
      "id": 76522,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV048179_dynmc",
      "rank": 236028,
      "technique": "Lithograph",
      "title": "Twentysix Gasoline Stations",
    },
    Object {
      "id": 77264,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:DDC107120_dynmc",
      "rank": 267319,
      "technique": "Etching",
      "title": "Arrival in Naples at the Piazza di Spagna",
    },
    Object {
      "id": 77304,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:DDC107121_dynmc",
      "rank": 267320,
      "technique": "Etching",
      "title": "Viceroy's Arrival at the Pleasure-Garden in Naples",
    },
    Object {
      "id": 77297,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:DDC107129_dynmc",
      "rank": 267328,
      "technique": "Etching",
      "title": "Horse-Drawn Carriage with Well-Dressed Figures and Bathers; Fortress and Port in the Background",
    },
    Object {
      "id": 77295,
      "primaryimageurl": "https://nrs.harvard.edu/urn-3:HUAM:DDC107153_dynmc",
      "rank": 267351,
      "technique": "Etching",
      "title": "Nonce Palace at San Francesco della Vigna in Venice",
    },
  ],
}
`;
